### PR TITLE
Palo alto networks ai ops

### PR DESCRIPTION
**Pull request by tcarmeli1**
## Status
- [x] Ready

## Related Issues
fixes: [link to the issue](https://jira-dc.paloaltonetworks.com/browse/CIAC-9040)

## Description
bpa polling integration

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps.py [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+    else:
+        try:
+            # Add xml tag to xml
+            xml_header = '<?xml version="1.0"?>'
+            result = f'{xml_header}\n {config_file}'
+            sio_xml = io.StringIO(result)
+            xml_in_bytes = sio_xml.read().encode()
+            return xml_in_bytes
+        except Exception:
+            raise DemistoException("The downloaded config file could not be converted.")
```
Comment (jbabazadeh) [inline Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps.py:229]:
```suggestion
            raise DemistoException("The downloaded config file from Panorama/Pan-os could not be converted.")
```

File: Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps.py [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+    except Exception:
+        raise DemistoException("Request Succeeded, A parse error occurred.")
+
+
+def get_values_from_xml(xml_string, tags):
+    try:
+        result = []
+        root = ET.fromstring(xml_string)
+        for tag in tags:
+            result.append(root.find(tag).text)  # type: ignore
```
Comment (jbabazadeh) [inline Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps.py:203]: what if `root.find(tag)` return `None`?

File: Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps.py [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+
+
+''' HELPER FUNCTIONS '''
+
+
+def adjust_xml_format(xml_string, new_root_tag):
+    try:
+        root = ET.fromstring(xml_string)
+        sub_tags = root.find(f'.//{new_root_tag}')
+        if sub_tags is not None:
```
Comment (jbabazadeh) [inline Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps.py:187]:
```suggestion
        if sub_tags:
```

File: Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps.py [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+            attributes = ' '.join([f'{k}="{v}"' for k, v in sub_tags.attrib.items()])
+        new_xml = f"<{new_root_tag} {attributes}>"
+        for child in sub_tags:  # type: ignore
+            new_xml += ET.tostring(child, encoding="unicode")
+        new_xml += f"</{new_root_tag}>"
+        return new_xml
+    except Exception:
+        raise DemistoException("Request Succeeded, A parse error occurred.")
```
Comment (jbabazadeh) [inline Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps.py:195]: please change the error message to be more informative

**Review by jbabazadeh — CHANGES_REQUESTED** [review]
Great job!!
a few comments.

File: Packs/BPA/pack_metadata.json [hunk 5]
```diff
@@ -1,5 +1,5 @@
 {
-    "name": "Best Practice Assessment (BPA) by Palo Alto Networks",
+    "name": "Best Practice Assessment (BPA) by Palo Alto Networks (Deprecated)",
```
[the remark this answers is not in the document]
Comment (tcarmeli1) [reply Packs/BPA/pack_metadata.json]: DID IT HERE: https://github.com/demisto/content/pull/33520

**Review by BEAdi — COMMENTED** [review]
Great job!!
Some minor comments

File: Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps.py [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+            raise DemistoException("Missing upload status, Error: parse Error.")
+        return status
+
+    def download_bpa_request(self, report_id):
+        headers = {
+            'Accept': 'application/json',
+            'Authorization': f'Bearer {self._access_token}'
+        }
+        res = self._http_request(method='GET',
+                                 full_url=f'https://api.stratacloud.paloaltonetworks.com/aiops/bpa/v1/reports/{report_id}',
```
[the remark this answers is not in the document]
Comment (tcarmeli1) [reply Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps.py:156]: Yes. talked in zoom :)

File: Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps.yml [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+description: Palo Alto Networks Best Practice Assessment (BPA) analyzes NGFW and Panorama configurations and compares them to the best practices.
+display: Palo Alto Networks AIOps
+name: PaloAltoNetworksAIOps
+script:
+  commands:
+  - name: aiops-bpa-report-generate
+    arguments:
+    - name: entry_id
+      description: |-
+        - Entry_id from xsoar war-room after uploading a file - should be a config file in xml format.
```
Comment (ShirleyDenkberg) [inline Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps.yml:50]:
```suggestion
        - Entry_id from Cortex XSOAR War Room after uploading a file - should be a config file in xml format.
```

File: Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps.yml [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    - name: interval_in_seconds
+      description: Interval for polling mechanism.
+      defaultValue: "30"
+    - name: timeout
+      default: true
+      description: Timeout for downloading the file.
+      defaultValue: "600"
+    - name: export_as_file
+      auto: PREDEFINED
+      description: Export the generated report as a file.
```
Comment (ShirleyDenkberg) [inline Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps.yml:67]:
```suggestion
      description: Whether to export the generated report as a file.
```

File: Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps.yml [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+      defaultValue: true
+      predefined: 
+      - 'true'
+      - 'false'
+    - name: show_in_context
+      auto: PREDEFINED
+      description: Show the report data inside the context.
```
Comment (ShirleyDenkberg) [inline Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps.yml:74]:
```suggestion
      description: Whether to show the report data inside the context.
```

File: Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps.yml [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+      defaultValue: false
+      predefined: 
+      - 'true'
+      - 'false'
+    description: |- 
+      Generates a bpa report. Steps - 
+      - Get information about Pan-OS/Panorama device.
+      - Get configuration file of Pan-OS/Panorama, if the user provides an entry_id to a config file this steps is skipped.
```
Comment (ShirleyDenkberg) [inline Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps.yml:82]:
```suggestion
      - Get configuration file of Pan-OS/Panorama. If the user provided an entry_id to a config file this step is skipped.
```

File: Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps.yml [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+      - Use the two above in order to generate a bpa report.
```
Comment (ShirleyDenkberg) [inline Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps.yml:83]:
```suggestion
      - Use the information retrieved above to generate a BPA report.
```

File: Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps.yml [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+      - During this process the API generates also a report_id for internal use.
```
Comment (ShirleyDenkberg) [inline Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps.yml:84]:
```suggestion
      - During this process the API also generates a report_id for internal use.
```

File: Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps.yml [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+      required: true
+    - name: hide_polling_output
+      description: Hide polling output from war-room.
+      required: true
+      defaultValue: true
+    - name: export_as_file
+      description: Export the generated report as a file.
+      defaultValue: true
+    - name: show_in_context
+      description: Show the report data inside the context.
```
Comment (ShirleyDenkberg) [inline Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps.yml:99]:
```suggestion
      description: Whether to export the generated report as a file.
      defaultValue: true
    - name: show_in_context
      description: Whether to show the report data inside the context.
```

File: Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps.yml [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+      defaultValue: false
+    - name: first_round
+      description: Does this polling round is the first round.
```
Comment (ShirleyDenkberg) [inline Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps.yml:102]:
```suggestion
      description: Whether this polling round is the first round.
```

File: Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps_description.md [hunk 15]
```diff
@@ -0,0 +1,27 @@
+## AIOps integration 
+
+- The Palo Alto Networks Best Practice Assessment (BPA) measures your usage of our Next-Generation Firewall (NGFW) and Panorama™ security management capabilities across your deployment, enabling you to make adjustments that maximize your return on investment and strengthen security.
+- This Integration enables you to programmatically generate BPA data for both the free and premium instances of AIOps for NGFW.
```
Comment (ShirleyDenkberg) [inline Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps_description.md:4]:
```suggestion
- This integration enables you to programmatically generate BPA data for both the free and premium instances of AIOps for NGFW.
```

File: Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps_description.md [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+
+## AIOps Authentication:
+- Follow the instructions in this [link](https://pan.dev/aiops-ngfw-bpa/api/) to create your own client_id, client_secret, tsg_id.
+- Use these client_id, client_secret, tsg_id to authenticate to your AIOps instance.
+### Panorama / Pan-OS Authentication:
+- Use your server url, username and password in order to get your API-key.
```
Comment (ShirleyDenkberg) [inline Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps_description.md:10]:
```suggestion
- Use your server URL, username and password in order to get your API-key.
```

File: Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps_description.md [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+```xml
+<response status="success">
+    <result>
+        <key>
+            gJlQWE56987nBxIqyfa62sZeRtYuIo2BgzEA9UOnlZBhU==
+        </key>
+    </result>
+</response>
+```
+- Optional: revoke all currently valid API keys, follow this [link](https://docs.paloaltonetworks.com/pan-os/9-1/pan-os-panorama-api/get-started-with-the-pan-os-xml-api/get-your-api-key)
```
Comment (ShirleyDenkberg) [inline Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps_description.md:27]:
```suggestion
- Optional: To revoke all currently valid API keys, follow this [link](https://docs.paloaltonetworks.com/pan-os/9-1/pan-os-panorama-api/get-started-with-the-pan-os-xml-api/get-your-api-key)
```

File: Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/README.md [hunk 18]
```diff
@@ -0,0 +1,53 @@
+Palo Alto Networks Best Practice Assessment (BPA) analyzes NGFW and Panorama configurations and compares them to the best practices.
+This integration was integrated and tested with version from March 2024 of PaloAltoNetworksAIOps.
```
Comment (ShirleyDenkberg) [inline Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/README.md:2]:
```suggestion
This integration was integrated and tested with the March 2024 version of PaloAltoNetworksAIOps.
```

File: Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/README.md [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+
+#### Base Command
+
+`pan-aiops-bpa-report-generate`
+
+#### Input
+
+| **Argument Name** | **Description** | **Required** |
+| --- | --- | --- |
+| entry_id | Entry_id from xsoar war-room after uploading a file. | Optional |
```
Comment (ShirleyDenkberg) [inline Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/README.md:39]:
```suggestion
| entry_id | Entry_id from Cortex XSOAR War Room after uploading a file. | Optional | 
```

File: Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/README.md [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+
+#### Context Output
+
+There is no context output for this command.
+#### Command example
+```!pan-aiops-bpa-report-generate requester_email=tal@gmail.com requester_name=tall```
+#### Human Readable Output
+
+>None
+
```
Comment (ShirleyDenkberg) [inline Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/README.md:53]: I think this README has to be regenerated as it is missing much information from the yml file.

**Comment by ShirleyDenkberg** [comment]
@BEAdi @jbabazadeh Doc review completed.
@tcarmeli1 Please regenerate the integration readme.

**Review by jbabazadeh — APPROVED** [review]
Amazing!

File: Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps_description.md [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+curl -H "Content-Type: application/x-www-form-urlencoded" 
+-X POST https://firewall/api/?type=keygen 
+-d 'user=<user>&password=<password>'
+```
+- A successful API call returns status="success" along with the API key within the key element:
+```xml
+<response status="success">
+    <result>
+        <key>
+            gJlQWE56987nBxIqyfa62sZeRtYuIo2BgzEA9UOnlZBhU==
```
[the remark this answers is not in the document]
Comment (tcarmeli1) [reply Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps_description.md:22]: What do you mean here?
changed the value to api-key 
thats what you mean?

File: Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps.yml [hunk 22]
```diff
@@ -0,0 +1,121 @@
...
+    - name: export_as_file
+      auto: PREDEFINED
+      description: Whether to export the generated report as a file.
+      defaultValue: true
+      predefined: 
+      - 'true'
+      - 'false'
+    - name: show_in_context
+      auto: PREDEFINED
+      description: Whether to show the report data inside the context.
```
[the remark this answers is not in the document]
Comment (tcarmeli1) [reply Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps.yml:77]: it does this automatically when using defaultValue (line below)

File: Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps.yml [hunk 23]
```diff
@@ -0,0 +1,121 @@
...
+    - name: interval_in_seconds
+      description: Interval for polling mechanism.
+      defaultValue: "30"
+    - name: timeout
+      default: true
+      description: Timeout for downloading the file.
+      defaultValue: "600"
+    - name: export_as_file
+      auto: PREDEFINED
+      description: Whether to export the generated report as a file.
```
[the remark this answers is not in the document]
Comment (tcarmeli1) [reply Packs/PaloAltoNetworksAIOps/Integrations/PaloAltoNetworksAIOps/PaloAltoNetworksAIOps.yml:70]: it does this automatically when using defaultValue (line below)

**Pull request merged**